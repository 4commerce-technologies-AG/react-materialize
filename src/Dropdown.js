import React, {
  Fragment,
  Children,
  cloneElement,
  useRef,
  useEffect
} from 'react';
import PropTypes from 'prop-types';
import idgen from './idgen';
import cx from 'classnames';

const Dropdown = ({ children, className, trigger, options, ...props }) => {
  const _dropdownContent = useRef(null);

  useEffect(() => {
    const instance = M.Dropdown.init(
      document.querySelector(`[data-target="${props.id}"]`),
      options
    );

    return () => {
      instance && instance.destroy();
    };
  }, [options, props.id, children]);

  const renderTrigger = () =>
    cloneElement(trigger, {
      'data-target': props.id,
      className: cx(trigger.props.className, 'dropdown-trigger')
    });

  const renderItems = () =>
    Children.map(children, element => {
      if (element.type.name === 'Divider') {
        return <li key={idgen()} className="divider" tabIndex="-1" />;
      } else {
        return <li key={idgen()}>{element}</li>;
      }
    });

  return (
    <Fragment>
      {renderTrigger()}
      <ul
        {...props}
        className={cx('dropdown-content', className)}
        ref={_dropdownContent}
      >
        {renderItems()}
      </ul>
    </Fragment>
  );
};

export const dropdownOptions = PropTypes.shape({
  alignment: PropTypes.oneOf(['left', 'right']),
  autoTrigger: PropTypes.bool,
  constrainWidth: PropTypes.bool,
  container: PropTypes.node,
  coverTrigger: PropTypes.bool,
  closeOnClick: PropTypes.bool,
  hover: PropTypes.bool,
  inDuration: PropTypes.number,
  outDuration: PropTypes.number,
  onOpenStart: PropTypes.func,
  onOpenEnd: PropTypes.func,
  onCloseStart: PropTypes.func,
  onCloseEnd: PropTypes.func
});

Dropdown.propTypes = {
  id: PropTypes.string,
  /**
   * The node to trigger the dropdown
   */
  trigger: PropTypes.node.isRequired,
  children: PropTypes.node,
  className: PropTypes.string,

  /**
   * Options hash for the dropdown
   * <a target="_blank" href="http://materializecss.com/dropdown.html#options">http://materializecss.com/dropdown.html</a>
   */
  options: dropdownOptions
};

Dropdown.defaultProps = {
  id: `Dropdown_${idgen()}`,
  options: {
    alignment: 'left',
    autoTrigger: true,
    constrainWidth: true,
    container: null,
    coverTrigger: true,
    closeOnClick: true,
    hover: false,
    inDuration: 150,
    outDuration: 250,
    onOpenStart: null,
    onOpenEnd: null,
    onCloseStart: null,
    onCloseEnd: null
  }
};

export default Dropdown;
