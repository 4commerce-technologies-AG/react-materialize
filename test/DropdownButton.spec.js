import React from 'react';
import { shallow } from 'enzyme';
import Dropdown from '../src/Dropdown';
import Divider from '../src/Divider';
import mocker from './helper/new-mocker';

describe('<Dropdown />', () => {
  let wrapper;

  test('renders', () => {
    wrapper = shallow(<Dropdown trigger={<button>Drop me!</button>} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('does not update ID reference with each render', () => {
    wrapper = shallow(<Dropdown trigger={<span>hi</span>} />);
    const { idx } = wrapper.instance();
    wrapper.render();
    expect(wrapper.instance().idx).toEqual(idx);
  });

  test('does correctly render dividers', () => {
    wrapper = shallow(
      <Dropdown trigger={<span>hi</span>}>
        <a href="#">test</a>
        <Divider />
        <a href="#">test</a>
      </Dropdown>
    );
    const { idx } = wrapper.instance();
    wrapper.render();
    expect(wrapper.contains(<li className="divider" tanIndex="-1" />));
  });

  test('deletes unwatend props', () => {
    const options = { hover: true };
    wrapper = shallow(<Dropdown trigger={<span>hi</span>} options={options} />);
    expect(wrapper.find('ul').props()).not.toMatchObject(options);
  });

  describe('initializes', () => {
    const dropdownInitMock = jest.fn();
    const dropdownInstanceDestroyMock = jest.fn();
    const dropdownMock = {
      init: (el, options) => {
        dropdownInitMock(options);
        return { destroy: dropdownInstanceDestroyMock };
      }
    };

    const restore = mocker('Dropdown', dropdownMock);

    beforeEach(() => {
      dropdownInitMock.mockClear();
      dropdownInstanceDestroyMock.mockClear();
    });

    afterAll(() => {
      restore();
    });

    test('handles options prop', () => {
      const options = { constrainWidth: true };
      wrapper = shallow(
        <Dropdown trigger={<button>Drop me!</button>} options={options} />
      );
      expect(dropdownInitMock).toHaveBeenCalledWith(options);
    });
  });
});
