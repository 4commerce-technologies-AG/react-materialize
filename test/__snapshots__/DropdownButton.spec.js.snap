// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown /> renders 1`] = `
<Fragment>
  <button
    className="dropdown-trigger"
    data-target="dropdownmockId"
  >
    Drop me!
  </button>
  <ul
    className="dropdown-content"
    id="dropdownmockId"
  />
</Fragment>
`;
